### PR TITLE
Implement TH_API functions from the set 1

### PR DESCRIPTION
**Pull request by 0mp**
This branch is **WIP**.

I'll be happy if you could look at the current state of my code and see if I implement those functions correctly.

Related to #47

### CODE REVIEW

File: torch/lib/THD/base/Tensor.hpp [hunk 1]
```diff
@@ -47,6 +47,59 @@ struct TensorScalarInterface : public Tensor {
   using scalar_type = real;
   virtual TensorScalarInterface& fill(scalar_type value) = 0;
   virtual TensorScalarInterface& add(const Tensor& source, scalar_type salar) = 0;
+
+  virtual THTensor *THTensor_(newWithSize1d)(long size0_) = 0;
```
Comment (jytug) [inline torch/lib/THD/base/Tensor.hpp]: Those ought to be constructors, not tensor methods. On which instance of `THTensor` will you call `newWithSize1d` to create a new one?

File: torch/lib/THD/base/Tensor.hpp [hunk 2]
```diff
@@ -47,6 +47,59 @@ struct TensorScalarInterface : public Tensor {
...
+  virtual THTensor *THTensor_(newWithStorage3d)(THStorage *storage_, ptrdiff_t storageOffset_, long size0_, long stride0_, long size1_, long stride1_, long size2_, long stride2_) = 0;
+  virtual THTensor *THTensor_(newWithStorage4d)(THStorage *storage_, ptrdiff_t storageOffset_, long size0_, long stride0_, long size1_, long stride1_, long size2_, long stride2_, long size3_, long stride3_) = 0;
+  virtual THTensor *THTensor_(newClone)(THTensor *self) = 0;
+  virtual THTensor *THTensor_(newContiguous)(THTensor *tensor) = 0;
+  virtual THTensor *THTensor_(newSelect)(THTensor *tensor, int dimension_, long sliceIndex_) = 0;
+  virtual THTensor *THTensor_(newNarrow)(THTensor *tensor, int dimension_, long firstIndex_, long size_) = 0;
+  virtual THTensor *THTensor_(newTranspose)(THTensor *tensor, int dimension1_, int dimension2_) = 0;
+  virtual THTensor *THTensor_(newUnfold)(THTensor *tensor, int dimension_, long size_, long step_) = 0;
+  virtual void THTensor_(narrow)(THTensor *self, THTensor *src, int dimension_, long firstIndex_, long size_) = 0;
```
Comment (jytug) [inline torch/lib/THD/base/Tensor.hpp]: We stick to the convention that if a method's return value is `void` or irrelevant, the method should return `*this` (so the return type will be `THTensor&`. Moreover, please change the naming of these methods. `tensor.THTensor_(narrow)` looks silly

File: torch/lib/THD/base/tensors/generic/THTensor.cpp [hunk 3]
```diff
@@ -132,9 +132,271 @@ auto THTensor<real>::add(const Tensor &source, scalar_type value) -> THTensor& {
...
 }
 
+#undef non_const_cast
+
 template<>
 thd::Type THTensor<real>::type() const {
   return thd::type_traits<real>::type;
 }
 
+// template<>
```
Comment (jytug) [inline torch/lib/THD/base/tensors/generic/THTensor.cpp]: All of these will probably just be a call of an appropriate `TH` function on `tensor`. With some skill, you might be able to `sed` it :-)

File: torch/lib/THD/master_worker/master/generic/THDTensor.cpp [hunk 4]
```diff
@@ -49,12 +49,12 @@ THLongStorage *THDTensor_(newStrideOf)(THDTensor *self)
   return stride;
 }
 
-void THDTensor_(setFlag)(THDTensor *self, const char flag)
+void THDTensor_(setFlag)(THDTensor *self, char flag)
```
Comment (jytug) [inline torch/lib/THD/master_worker/master/generic/THDTensor.cpp]: Why so?

File: torch/lib/THD/master_worker/master/generic/THDTensor.cpp [hunk 5]
```diff
@@ -49,12 +49,12 @@ THLongStorage *THDTensor_(newStrideOf)(THDTensor *self)
...
 {
   self->flag |= flag;
 }
 
-void THDTensor_(clearFlag)(THDTensor *self, const char flag)
+void THDTensor_(clearFlag)(THDTensor *self, char flag)
```
Comment (jytug) [inline torch/lib/THD/master_worker/master/generic/THDTensor.cpp]: Why so?

File: torch/lib/THD/master_worker/worker/Dispatch.cpp [hunk 6]
```diff
@@ -1,6 +1,5 @@
 #include <TH/THStorage.h>
 #include <cstdint>
-#include <unordered_map>
```
Comment (jytug) [inline torch/lib/THD/master_worker/worker/Dispatch.cpp]: I think we need this here

File: torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp [hunk 7]
```diff
@@ -30,3 +29,263 @@ static void free(rpc::RPCMessage& raw_message) {
   (void)workerTensors.erase(tensor_id);
 }
 
+static void tensorNewWithSize1d(rpc::RPCMessage& raw_message) {
```
Comment (jytug) [inline torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp]: If you choose the `tensorNewWith...` convention over `tensorConstructWith`, please rename all the `construct` functions appropriately, including the `storage` ones. I names them `construct`, as `new` is a keyword, but that's not a problem anymore

**Review by jytug — CHANGES_REQUESTED** [review]
Remember we are in the C++ world now. Return references, not pointers, rename the class methods. Moreover, there are no changes in `master_worker/master/generic/THDTensor.cpp (.h)`. Nor in `master_worker/worker/dispatch/Tensor.cpp`, which are actually your main targets

File: torch/lib/THD/master_worker/master/generic/THDTensor.cpp (hunk 5 above)
Comment (0mp) [reply torch/lib/THD/master_worker/master/generic/THDTensor.cpp]: It is copied by value so there's no point in making it constant. At least, that's what they thought me in FreeBSD :D

File: torch/lib/THD/master_worker/worker/Dispatch.cpp (hunk 6 above)
Comment (0mp) [reply torch/lib/THD/master_worker/worker/Dispatch.cpp]: This library is included twice in this file.

**Pull request closed without merging**